### ISSUE TITLE
FileUpload: Did not receive any data in the allotted time.

### ISSUE THREAD

**Issue by Liero** [dotnet/aspnetcore#38842]
### Describe the bug

Not sure if this is a bug but I haven't found any documentation either.

   `IBrowserFile.ReportFile!.OpenReadStream(maxFileSize).CopyToAsync(fileStream);` sometimes throws 

> System.TimeoutException: Did not receive any data in the allotted time.

### To Reproduce

I haven't found a reliable way to reproduce the issue, since it happens only sometimes and have no idea what is causing it.

Here is the code:

```
IBrowserFile _selectedFile;

void OnSelectedFile(InputFileChangeEventArgs args)
{
    _selectedFile = args.File;
}

private async Task UploadMethod()
{
    string basePath = @"\\fileshare\folder";
    string relativePath = @"{MyFolder\DateTime.Now:dd.MM.yyyy}.pdf";
    string fullPath = Path.Combine(basePath, relativePath);
    using FileStream fs = new FileStream(fullPath, FileMode.Create);

    const long maxFileSize = 5000000; // 5 MB
    if (_selectedFile.Size > maxFileSize)
    {
        Toasts.ShowError("Chyba pri ukladaní. Súbor OTDR má viac ako 5 MB.", "OTDR Report");
    }
    else
    {
        _selectedFile.OpenReadStream(maxFileSize).CopyToAsync(fs);
    }
}
```

### Exceptions (if any)

<details>
  <summary>System.TimeoutException: Did not receive any data in the allotted time. </summary>
```
[
   {
      "method":"System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw",
      "level":0,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.IO.Pipelines.Pipe.GetReadResult",
      "level":1,
      "line":0,
      "assembly":"System.IO.Pipelines, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51"
   },
   {
      "method":"System.IO.Pipelines.Pipe.GetReadAsyncResult",
      "level":2,
      "line":0,
      "assembly":"System.IO.Pipelines, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51"
   },
   {
      "method":"System.IO.Pipelines.PipeReaderStream+<ReadAsyncInternal>d__30.MoveNext",
      "level":3,
      "line":0,
      "assembly":"System.IO.Pipelines, Version=6.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51"
   },
   {
      "method":"System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw",
      "level":4,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess",
      "level":5,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification",
      "level":6,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSDataStream+<ReadAsync>d__36.MoveNext",
      "level":7,
      "line":0,
      "assembly":"Microsoft.AspNetCore.Components.Server, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60"
   },
   {
      "method":"System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw",
      "level":8,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess",
      "level":9,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification",
      "level":10,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"Microsoft.AspNetCore.Components.Forms.BrowserFileStream+<CopyFileDataIntoBuffer>d__30.MoveNext",
      "level":11,
      "line":0,
      "assembly":"Microsoft.AspNetCore.Components.Web, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60"
   },
   {
      "method":"System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw",
      "level":12,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess",
      "level":13,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification",
      "level":14,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"Microsoft.AspNetCore.Components.Forms.BrowserFileStream+<ReadAsync>d__28.MoveNext",
      "level":15,
      "line":0,
      "assembly":"Microsoft.AspNetCore.Components.Web, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60"
   },
   {
      "method":"System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw",
      "level":16,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess",
      "level":17,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification",
      "level":18,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.IO.Stream+<<CopyToAsync>g__Core|29_0>d.MoveNext",
      "level":19,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw",
      "level":20,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess",
      "level":21,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification",
      "level":22,
      "line":0,
      "assembly":"System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e"
   },
   {
      "method":"MyApp.Pages.MyPage+<UploadMethod>d__54.MoveNext",
      "level":23,
      "line":348,
      "assembly":"MyApp, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null",
      "fileName":"D:\\a\\1\\s\\src\\app\\Pages\\MyPage.razor.cs"
   }
]
```
</details>
</body>
</html>

### Further technical details
- ASP.NET Core version: 6
- Include the output of `dotnet --info`:

<details>
  <summary>dotnet --info Output</summary>

```
.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa
```

</details>

**Comment by TanayParikh** [comment]
This exception is raised in Blazor Server ([here](https://cs.github.com/dotnet/aspnetcore/blob/5adbdcb18ce1167321e919b930d575cbbfb7512c/src/Components/Server/src/Circuits/RemoteJSDataStream.cs?q=Did%20not%20receive%20any%20data%20in%20the%20allotted%20time#L211)) when the client fails to send any data chunks within the stream for the [SignalR `JSInteropDefaultCallTimeout`](https://cs.github.com/dotnet/aspnetcore/blob/5adbdcb18ce1167321e919b930d575cbbfb7512c/src/Components/Server/src/Circuits/RemoteJSRuntime.cs#L192).

Are you seeing the client send the stream data chunks within the Developer Tools > Network tab in the browser? Are you behind a VPN or proxy which may be restricting certain packets? If you can provide a public GitHub repro application, it'll allow us to diagnose the issue further.

**Comment by Jesus-QC** [comment]
Im having the same issue when I upload big files to the server, it does not happen with small ones.

**Comment by TanayParikh** [comment]
> Im having the same issue when I upload big files to the server, it does not happen with small ones.

This exception is raised in Blazor Server ([here](https://cs.github.com/dotnet/aspnetcore/blob/5adbdcb18ce1167321e919b930d575cbbfb7512c/src/Components/Server/src/Circuits/RemoteJSDataStream.cs?q=Did%20not%20receive%20any%20data%20in%20the%20allotted%20time#L211)) when the client fails to send any data chunks within the stream for the [SignalR `JSInteropDefaultCallTimeout`](https://cs.github.com/dotnet/aspnetcore/blob/5adbdcb18ce1167321e919b930d575cbbfb7512c/src/Components/Server/src/Circuits/RemoteJSRuntime.cs#L192).

Are you seeing the client send the stream data chunks within the Developer Tools > Network tab in the browser? Are you behind a VPN or proxy which may be restricting certain packets? If you can provide a public GitHub minimal repro application, it'll allow us to diagnose the issue further.

**Comment by Jesus-QC** [comment]
I just found out this:
![image](https://user-images.githubusercontent.com/69375249/147163017-8d407253-4846-4bf9-a199-273757a9b19b.png)
Let me try to fix that and i will send here the details.

**Comment by Jesus-QC** [comment]
Not happening anymore as far as I see!

**Comment by angularsen** [comment]
I don't have a repro, but we faced the same problem after upgrading Blazor server-side app from .NET5 to .NET6.

We could read the stream immediately after calling `OpenReadStream()`, but it hung when reading the stream in the Save button click event handler.

### Tried
- Specify max file size argument, or use the default (500kB).
- Calling `InvokeAsync` to ensure it was running on the main thread.
- Different file sizes, ~40 kB was consistently successful while 70+ kB was not.

### Workaround
Read and buffer the entire stream in `<InputFile OnChange="@LoadFiles" />` event handler, immediately after calling `OpenReadStream()`. Then we reused that buffer in the Save button click event handler. 

We lose the benefit of streaming, but in our use case these were small files and an internal tool, so it was acceptable.

### Notes
One interesting symptom was that all button click event handlers stopped responding once this hang occurred, and even a page reload did not fix it. We had to restart the Blazor server to fix it. This page required being logged in, so it might have been this particular circuit that had a deadlock of some sort. I really don't understand what was going on.

**Comment by TanayParikh** [comment]
@angularsen can you confirm if you're on Websockets (not long polling). You should see a warning message like [this](https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-999922601) if you're using long polling. 

Also, could you please follow the debug steps/question in my original post: https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-986905123

**Comment by angularsen** [comment]
@TanayParikh Yes, we use websockets.

```
[2022-01-20T07:46:19.250Z] Information: Normalizing '_blazor' to 'https://xxx/_blazor'. blazor.server.js:1 
[2022-01-20T07:46:19.422Z] Information: WebSocket connected to wss://xxx/_blazor?id=xxx
```

> Are you seeing the client send the stream data chunks within the Developer Tools > Network tab in the browser? 

Yes, binary messages are listed on the `_blazor` blocked request in Chrome dev tools while using the app. 

If I select a 40 kB file, I see its binary message listed. I can then click upload and it succeeds.

![image](https://user-images.githubusercontent.com/787816/150299831-4e26b6fc-8d1c-4185-aead-386242f456ea.png)

If I select 80 kB file, I see 3 chunked messages of ~32 kB with its contents. However, the UI now hangs and nothing happens and no additional binary messages are logged when I click the upload button, until it times out and crashes the app after 30 seconds.

![image](https://user-images.githubusercontent.com/787816/150299165-76cd2ece-a125-44b4-ac07-1415867c01c9.png)

![image](https://user-images.githubusercontent.com/787816/150298089-aee24278-b222-4ee7-81d8-f42ccf1dda9b.png)

In my debugging, it seems reading from the file stream is the root cause. At this point, the blazor app has crashed and is no longer responding.

> Are you behind a VPN or proxy which may be restricting certain packets?

No.

> If you can provide a public GitHub minimal repro application, it'll allow us to diagnose the issue further.

I will try to find the time, but I can't promise anything.

**Comment by TanayParikh** [comment]
Thanks @angularsen, that's some great data. Is this a flaky issue, or is it 100% reproducible within your current application? If you haven't already, can you please increase your log levels to Debug/Verbose to ensure we can capture the most data. Setting `DetailedErrors: true` (per the console log message in your screenshot) may also be helpful to capture the SignalR logs.

>> If you can provide a public GitHub minimal repro application, it'll allow us to diagnose the issue further.
> I will try to find the time, but I can't promise anything.

Thanks, this'll be optimal and ensure I'm able to appropriately investigate this issue.

I'm re-opening this issue ticket.

**Comment by angularsen** [comment]
It is 100% consistent in our app. Thanks for reopening.

**Comment by Liero** [comment]
@TanayParikh : I confirm that the symptoms described by @angularsen the same as we experienced. In my original post, the stream is opened fo reading in button click event callback, however I was not able to reproduce it. just found the exception in logs.

I suspect that the key is the delay between Selecting the file and reading the stream

**Comment by andrzejczar84** [comment]
Hello, same issue here. 
Blazor server side. Working well at .Net5, failing on OpenReadStream().ReadAsync() at .Net6. 
We use this to upload *.jpg file to sql database.

Method launched as input file change for simple <input file... /> dialog.
```
        public int readbytes;
        protected async Task OnInputFileChange(InputFileChangeEventArgs e)
        {

            var format = "image/png";

            var imageFile = e.File;
            var resizedImageFile = await imageFile.RequestImageFileAsync(format,
                    400, 200);
            var buffer = new byte[resizedImageFile.Size];
            readbytes = await resizedImageFile.OpenReadStream().ReadAsync(buffer);
            imageDataUrl =
                $"data:{format};base64,{Convert.ToBase64String(buffer)}";

            ImgUploaded = buffer;

        }
```
In general byte array size is equal to file size. The problem begins in this line 

`readbytes = await resizedImageFile.OpenReadStream().ReadAsync(buffer);`

it reads only part of the buffer, displays part of the picture, then after several seconds throw error
 
![image](https://user-images.githubusercontent.com/34892451/150995289-6ab28043-1614-40df-a6e7-7a4628548979.png)
![image](https://user-images.githubusercontent.com/34892451/150995372-153662f3-909c-442d-badd-e5e3d41ddcb3.png)

What I investigated checking variable readbytes, it is always the same quantity for the same file. 
For example for this picture:
![miejsce wyst, defektu na czesci 400x200](https://user-images.githubusercontent.com/34892451/150995937-e14e0e1e-23e4-4bbc-86d1-1e827ec9b298.jpg)
It reads all the time 32256 bytes. Checked on MS Edge and Chrome. The same 32256bytes, so it seams to be not a connection problem.

What is more we have this problem on built soultion already deployed at the Windows Server, and on our VS2022 solutions in debug mode, and at all places in solution where we upload file stream this way.  Seams not to be a problem of single machine configuration.

**Comment by TanayParikh** [comment]
Thanks for the details @andrzejczar84. Would you be able to put together a minimal, public, GitHub repo that reproduces the issue so I may take a look. Also could you provide the image at an external host (GitHub compresses the image p as dyed into md I believe).

**Comment by andrzejczar84** [comment]
@TanayParikh here is the picture: https://imgur.com/a/q8L0jwj

I'll try to reproduce the problem at the new minimal repo as fas as it's possible

**Comment by andrzejczar84** [comment]
@TanayParikh Done.

Here is simple Blazor App reproducing problem: https://github.com/andrzejczar84/FileUploadBug.Net6

Screenshot below - the same jpg, the same bytes read, the same freez of app after several seconds

![image](https://user-images.githubusercontent.com/34892451/151012711-a8e01bb0-4c58-4f3a-92ac-5d7b4e8cda82.png)

**Comment by TanayParikh** [comment]
Thanks @andrzejczar84, I'm able to repro.

> I have checked also similar code to upload PDF files. Same way file stream read-only part of the file. 

Would you mind committing this into your github repo (omitting the DB dependency). 

Side note, I recommend checking out the new image upload guidance available [here](https://docs.microsoft.com/aspnet/core/blazor/images?view=aspnetcore-6.0) to help avoid having to Base64 encode images.

**Comment by andrzejczar84** [comment]
@TanayParikh repo updated with pdf upload.

I tried to upload 1,8Mb file.
![image](https://user-images.githubusercontent.com/34892451/151034712-f561fba1-e271-4f2b-b73d-514fb35f883f.png)

Method passed but uploaded only ~64kb
![image](https://user-images.githubusercontent.com/34892451/151034842-4d997600-394d-4616-ba04-ce616e956e97.png)

**Comment by andrzejczar84** [comment]
> Side note, I recommend checking out the new image upload guidance available [here](https://docs.microsoft.com/aspnet/core/blazor/images?view=aspnetcore-6.0) to help avoid having to Base64 encode images.

I tried it but the uploaded file doesn't appear in img tag after upload. I'm not familiar with using javascript in blazor, maybe You see what is missing here... I'm impressed that getting images from file dialog can be as far complicated. I'd love to see some "blackbox" which takes a file by file input, paste or drag and drop, and do something (I don't care how) to give me byte[] I can insert into sql database, and source for img tag.

```
@inject IJSRuntime JSRuntime
@inject System.Net.Http.HttpClient HttpClient

<img id="showImageHere" />

<InputFile OnChange="ResizeAndDisplayImageUsingStreaming" />     

@code {
    private async Task ResizeAndDisplayImageUsingStreaming(InputFileChangeEventArgs e)
    {
        var imageFile = e.File;
        var resizedImage = 
            await imageFile.RequestImageFileAsync("image/jpg", 250, 250);
        var jsImageStream = resizedImage.OpenReadStream();
        var dotnetImageStream = new DotNetStreamReference(jsImageStream);
        await JSRuntime.InvokeVoidAsync("setImageUsingStreaming", 
            "showImageHere", dotnetImageStream);
    }
    private async Task SetImageUsingStreamingAsync()
    {
        var imageStream = await GetImageStreamAsync();
        var dotnetImageStream = new DotNetStreamReference(imageStream);
        await JSRuntime.InvokeVoidAsync("setImageUsingStreaming", 
            "image1", dotnetImageStream);
    }
    private async Task<Stream> GetImageStreamAsync()
    {
        return await HttpClient.GetStreamAsync(
            "https://avatars.githubusercontent.com/u/9141961");
    }
}
```

**Comment by huseyin-ezemi** [comment]
Hi to all
this code worked for me : 

   ```
                    var buffer = new byte[e.File.Size];
                    var fileContent = new StreamContent(e.File.OpenReadStream(1024 * 1024 * 15));
                    buffer = await fileContent.ReadAsByteArrayAsync();
```

**Comment by Bobetti** [comment]
After upgrading my Blazor Server Side application, I found out that file upload to Azure blob is not working anymore for any file.

For example, if I upload 5MB image file, I get on azure blob a 50MB ruined image file. If I upload a video file 12Mb, I get on azure 120MB ruined file.

Here is my upload code:

```
 private async Task<UploadResult> UploadFileInChunksAsync(string prefix, IBrowserFile file, CancellationToken cancellationToken)
        {

            var result = new UploadResult()
            {
                FileName = file.Name
            };

            try
            {

                var fullFileName = string.Format("{0}{1}", prefix, file.Name);

                var blockBlobClient = _containerClient.GetBlockBlobClient(fullFileName);

                int blockSize = 1 * 1024 * 1024; //1 MB Block
                int offset = 0;
                int counter = 0;

                List<string> blockIds = new List<string>();

                using (var fs = file.OpenReadStream(_maxAllowedSize))
                {
                    var bytesRemaining = fs.Length;

                    do
                    {
                        var dataToRead = Math.Min(bytesRemaining, blockSize);
                        byte[] data = new byte[dataToRead];
                        var dataRead = await fs.ReadAsync(data, offset, (int)dataToRead);
                        bytesRemaining -= dataRead;
                        if (dataRead > 0)
                        {
                            var blockId = Convert.ToBase64String(Encoding.UTF8.GetBytes(counter.ToString("d6")));
                            await blockBlobClient.StageBlockAsync(blockId, new MemoryStream(data));
                            Console.WriteLine(string.Format("Block {0} uploaded successfully.", counter.ToString("d6")));
                            blockIds.Add(blockId);
                            counter++;

                            cancellationToken.ThrowIfCancellationRequested();
                        }
                    }
                    while (bytesRemaining > 0);

                    // Console.WriteLine("All blocks uploaded. Now committing block list.");

                    await blockBlobClient.CommitBlockListAsync(blockIds);

                    // Console.WriteLine("Blob uploaded successfully!");

                    result.Uploaded = true;
                    result.Uploading = false;
                }
            }

            catch (Exception ex)
            {
                result.Uploaded = false;
                result.ErrorMessage = ex.Message;
            }

            return result;
        }
```
After digging, I found out that this line:
 var dataRead = await fs.ReadAsync(data, offset, (int)dataToRead);
always reads only 32256 bytes of info

Here are some test results with different solutions: .Net5 and .Net6 - uploading the same file:

correctly working upload in .NET 5
var bytesRemaining = fs.Length => 5317702
dataToRead => 1048576
dataRead => 1048576
bytesRemaining => 4269126
.
.
.

5 junks in total

=====

Not working .NET 6 results

bytesRemaining => 5317702
dataToRead => 1048576
**dataRead => 32256**

50 junks in total, ruined file

Any help with what to do next?

**Comment by Bobetti** [comment]
After some thinking and considering that in .Net 6 ReadAsync function reads the random number of data and not like in .Net 5, specific amount, I rewrote upload function this way:

```
 int blockSize = 1 * 1024 * 1024; //1 MB Block         
 var bytesRead = 0;
                var buffer = new byte[blockSize];

                while ((bytesRead = await fs.ReadAsync(buffer)) != 0)
                {
                    using var writeStream = new MemoryStream();
                    await writeStream.WriteAsync(buffer, 0, bytesRead);
                    writeStream.Position = 0;

                    var blockId = Convert.ToBase64String(Encoding.UTF8.GetBytes(counter.ToString("d6")));
                    await blockBlobClient.StageBlockAsync(blockId, writeStream);                  

                    counter++;

                    blockIds.Add(blockId);

                }

                await blockBlobClient.CommitBlockListAsync(blockIds);
```

Now, seems to me everything works fine. Files uploaded in correct size.

**Comment by TanayParikh** [comment]
Hi all, I took another look through the repro, and that behavior is by design.

Please note per the [`Stream.ReadAsync` docs](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.readasync?view=net-6.0):

> The result value can be less than the number of bytes allocated in the buffer if that many bytes are not currently available, or it can be 0 (zero) if the end of the stream has been reached.

That's essentially what's happening in this case. When a single call is made to `ReadAsync`, it only gives you the first (few) chunks that have been sent from JS, you are not guaranteed to receive the full file. Blazor subsequently [throws an error](https://cs.github.com/dotnet/aspnetcore/blob/5adbdcb18ce1167321e919b930d575cbbfb7512c/src/Components/Server/src/Circuits/RemoteJSDataStream.cs?q=Did%20not%20receive%20any%20data%20in%20the%20allotted%20time#L211) as you've only read a part of the stream and the stream subsequently times out.

The way you'd want to achieve this would be by leveraging [`Stream.CopyToAsync`](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copytoasync?view=net-6.0) or by using `Stream.ReadAsync` using smaller chunks until you reach the end of the stream.  Note this is in line with our [streaming](https://docs.microsoft.com/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript?view=aspnetcore-6.0#stream-from-javascript-to-net) and [file upload](https://docs.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-6.0&pivots=server) guidance. Specific examples and sample code are also available in those docs pages.

Note in @huseyin-ezemi's [comment](https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-1043976343) above, they used `ReadAsByteArrayAsync`, that could also work as they're allowing the whole stream to be read. Note this may pose some security & scalability concerns as you're storing the whole stream/file in memory. This approach isn't recommended.

```
                 var fileContent = new StreamContent(e.File.OpenReadStream(1024 * 1024 * 15));
                 buffer = await fileContent.ReadAsByteArrayAsync();
```

The above explains the behavior presented within @andrzejczar84's repro app, however in the original issue @Liero mentions running into this with `CopyToAsync`. This is more concerning, however I've been unable to repro. If @Liero, @angularsen  or anyone else has a repro of this issue with `CopyToAsync`, I'd very much be interested in it. I'm marking this issue as "needs author feedback", it'll get auto-closed after a while if a repro can't be found. Though please feel free to re-open / start a new issue (and link this issue) if you do end up finding a repro with `CopyToAsync`.

**Comment by TanayParikh** [comment]
From @Liero's comment above:

> I confirm that the symptoms described by @angularsen the same as we experienced. In my original post, the stream is opened fo reading in button click event callback, however I was not able to reproduce it. just found the exception in logs.
>
> I suspect that the key is the delay between Selecting the file and reading the stream

I think you're onto something here. The timeout for the stream begins when you call [`OpenReadStream`](https://github.com/dotnet/aspnetcore/blob/9ad5dc9dccfdec2029eaf85dcc913ccc8c393610/src/Components/Web/src/Forms/InputFile/BrowserFile.cs#L36) which calls into the [`BrowserFileStream`](https://github.com/dotnet/aspnetcore/blob/9ad5dc9dccfdec2029eaf85dcc913ccc8c393610/src/Components/Web/src/Forms/InputFile/BrowserFileStream.cs#L86-L99) which calls into the [stream reference](https://github.com/dotnet/aspnetcore/blob/989e5756554cceb20b9a082049eeb421a7819619/src/JSInterop/Microsoft.JSInterop/src/Implementation/JSStreamReference.cs#L41), leading to the [runtime](https://github.com/dotnet/aspnetcore/blob/989e5756554cceb20b9a082049eeb421a7819619/src/Components/Server/src/Circuits/RemoteJSRuntime.cs#L192) which in turn kicks of the [stream and timeout](https://github.com/dotnet/aspnetcore/blob/989e5756554cceb20b9a082049eeb421a7819619/src/Components/Server/src/Circuits/RemoteJSDataStream.cs#L74).

Hence, if you're doing `OpenReadStream` and then waiting (past the timeout) to actually read from the stream, you'd encounter this error.

**Comment by Liero** [comment]
@TanayParikh: I've later realized, that the problem only occurred when the app was running in on-premise IIS, which has websockets disabled by default.

So in order to reproduce it:
1. use long polling instead of websockets 
2. select larger file to upload (5MB)
3. call `IBrowserFile.OpenReadAsync` from click or form submit callback one minute later after `InputFile.Change` callback (see original post)

**Comment by hri123phapale** [comment]
I think I do have same issue with .Net6 when migrated from .Net5.

Below mention code was perfectly working with .Net5 but recently we have migrated to .Net6 and file uploading stop working with the files having size more that 400-500 KB.
Its working fine with small size files.
If we uncomment the commented lines(2 lines) from below code , its start working again.

protected async Task OnChangeHandler(InputFileChangeEventArgs args)
     {
         IReadOnlyList<IBrowserFile> files = args.GetMultipleFiles(100);
         List<UploadItem> items = new List<UploadItem>();
         int currentCount = 0;
         foreach (var file in files)
         {
             if (!IsValidFileType(file))
             {
                 continue;
             }
             if (currentCount >= CurrentAllowedAmount)
             {
                 break;
             }
             currentCount++;
             var fileStream = file.OpenReadStream(_maxFileSize * ByteToMByte);
             //byte[] buffer = new byte[file.Size];
             var fileContent = new StreamContent(fileStream);
             //buffer = await fileContent.ReadAsByteArrayAsync();
             items.Add(new UploadItem
             {
                 FileContent = fileStream,
                 Content = fileContent, 
                 File = file,
                 State = UploadState.ToUpload,
                 MaxFileSize = MaxFileSize,
                 IsFileValid = IsFileValid(file)
             });
         }
         await OnChange.InvokeAsync(items);
         Message = GetLatestMessage();
     }

**Comment by TanayParikh** [comment]
@hri123phapale please see https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-1151415351 above. I believe this is likely the same thing happening here.

**Comment by Liero** [comment]
@hri123phapale: Also make sure, websockets are enabled, see 
> ![image](https://user-images.githubusercontent.com/69375249/147163017-8d407253-4846-4bf9-a199-273757a9b19b.png)

**Comment by hri123phapale** [comment]
@Liero Yes, it connected
![image](https://user-images.githubusercontent.com/98818733/187620277-8a19de92-628f-4b7a-9e9f-3a755282d4ca.png)

**Comment by Liero** [comment]
hey, I see that you open the stream without reading it:

```csharp
var fileStream = file.OpenReadStream(_maxFileSize * ByteToMByte);
var fileContent = new StreamContent(fileStream);
```
and pass it to evencallback:

```csharp
items.Add(new UploadItem
{
  FileContent = fileStream,
  Content = fileContent, 
  File = file,
}
await OnChange.InvokeAsync(items);


so you attempt to read it later in the event handler probably. This is wrong. You have to read the stream immediately after calling `OpenReadStream()`, otherwise it can timeout.

**Comment by hri123phapale** [comment]
@Liero , Same code was perfectly working in .NET5. We have migrated to .NET6. But same code is not working with .NET6.

**Comment by Liero** [comment]
> Same code was perfectly working in .NET5
I'm not saying otherwise. I'm saying you should follow the guidelines.

**Comment by Liero** [comment]
@TanayParikh:
maybe it would be worth updating the [blazor upload guideline](https://docs.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-6.0&pivots=server) with https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-1151415351 - the warning about ReadBytes() and  about websockets.

BTW, using Long polling, there is probably no way to upload larger file, it is?

**Comment by hrvoje-grabusic** [comment]
@Liero Also in those docs Pages/FileUpload2.razor example code causes exactly this issue when uploading more then one file.
When uploading two files it doesn't read the stream right away and it freezes.

**Comment by Bluejanis** [comment]
We are using IIS, but have websockets enabled and still experience this error. It only occurs when uploading 2 or more files simultaneously, which are at least 400kb of size.

3 Files of 200kb => no issue
1 File of 8MB => no issue
2 Files of 400kb => System.TimeoutException: 'Did not receive any data in the allotted time.'

**Comment by sipi41** [comment]
I'm having also this problem... what I did is to read the File Input content using Javascript, in other words have the BLOB data transfered via JS and then it did work but certainly, a random issue with MS inputfile component.

`
Stream stream = file.OpenReadStream(maxAllowedSize: 2147483648);

                    MemoryStream memoryStream = new MemoryStream();

                    await stream.CopyToAsync(memoryStream); //----> PROBLEMATIC LINE`

**Comment by BhavanaShrotri** [comment]
I am facing the same issue after migrating Dotnet 5 to 6.
![image](https://user-images.githubusercontent.com/66585060/199690173-2c4538a6-5c99-4a25-8efb-67a0f9e6b050.png)

**Comment by oruchreis** [comment]
Has any update for this issue? I've tried with many browsers and with or without websocket connected. I've also tried with net 7, it still can't read the stream and throws `Did not receive any data in the allotted time`. So this error occurs after upgrading to .net 6 and .net 7, and it doesn't occur in .net 5.

**Comment by TanayParikh** [comment]
We've investigated and verified this bug. We're planning on tackling this in the next release.

**Comment by qw-34** [comment]
We have same issue, but it works on .NET 6 and only broken after migrate to .NET 7.

**Comment by haas-daniel** [comment]
We had the same issue in .NET 7 and fixed it by setting the SignalR DisableImplicitFromServicesParameters-option to true:
https://learn.microsoft.com/en-us/dotnet/api/Microsoft.AspNetCore.SignalR.HubOptions.DisableImplicitFromServicesParameters?view=aspnetcore-7.0&viewFallbackFrom=net-7.0

```
builder.Services
    .AddServerSideBlazor()
    .AddHubOptions(opt =>
    {
        opt.DisableImplicitFromServicesParameters = true;
    });
```

Background:
SignalR tries to call this method https://github.com/dotnet/aspnetcore/blob/de68feabb4dc84d2d5fbcfd7b3ed0b08f53216fa/src/Components/Server/src/ComponentHub.cs#L225 but fails because of incorrect argument count.

This method is registered in SignalR with 3 parameters instead of 4. That's because we use Autofac as IoC container and Autofac tells SignalR that it can provide a service for the byte[] parameter.
https://github.com/dotnet/aspnetcore/blob/971df4cbdc045c879f288b8cdfd726fd4212f882/src/SignalR/server/Core/src/Internal/HubMethodDescriptor.cs#L83

**Comment by DavidWaterfield** [comment]
@haas-daniel That solution worked for us. Thanks for sharing

**Comment by oruchreis** [comment]
I'm using DryIoc with `builder.Host.UseServiceProviderFactory`, and I was faced this issue. After I saw that uninstalling the DryIoc library fixed it, I reported it to the DryIoc repo here: https://github.com/dadhi/DryIoc/issues/547
But I don't know which repo should handle this issue. Should this bug be fixed by DryIoc and other DI libraries like Autofac or is it a topic here?

**Comment by 12evenant** [comment]
Hi All, In my case I was able to solve this by closing the stream.

```
    private async Task<bool> Upload(IBrowserFile file)
    {
        var fileBytes = new byte[file.Size];
        var stream = file.OpenReadStream();
        try
        {
            await stream.ReadAsync(fileBytes);
            //do something
            return true;
        }
        catch (Exception ex)
        {
            //do something
            return false;
        }
        finally
        {
            stream.Close();
        }
    }
```

**Comment by robi26** [comment]
Having exactly the same issue. We are using Lamar. @haas-daniel thanks for the workaround!

**Comment by alpisala02** [comment]
We have same issue after updating .NET 6 to .NET 7 on InputFile component. Unfortunately the workarounds didnt work for us.

![Screenshot 2023-01-26 105343](https://user-images.githubusercontent.com/71903661/214783960-728ab59c-8328-47a0-be82-a52de3fd6293.png)

**Comment by frutasAka** [comment]
I have the same issue and the workarounds didnt work. Any other solution?

**Comment by yandoldonov** [comment]
i am having the same issue calling calling this combination:

```
await using FileStream fs = new(path, FileMode.Create);
await file.OpenReadStream(maxFileSize).CopyToAsync(fs);
```

Stack Trace:

```
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at System.Runtime.CompilerServices.ConfiguredValueTaskAwaitable`1.ConfiguredValueTaskAwaiter.GetResult()
   at System.IO.Pipelines.PipeReaderStream.<ReadAsyncInternal>d__30.MoveNext()
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Components.Server.Circuits.RemoteJSDataStream.<ReadAsync>d__36.MoveNext()
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Components.Forms.BrowserFileStream.<CopyFileDataIntoBuffer>d__30.MoveNext()
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult()
   at Microsoft.AspNetCore.Components.Forms.BrowserFileStream.<ReadAsync>d__28.MoveNext()
   at System.IO.Stream.<<CopyToAsync>g__Core|27_0>d.MoveNext()
```

is there any progress on this?

**Comment by frutasAka** [comment]
> I have the same issue and the workarounds didnt work. Any other solution?

Since I was using ABP, I need to use @haas-daniel solution this way

`
        Configure<HubOptions>(options =>
        {
            options.DisableImplicitFromServicesParameters = true;
        });
`

**Comment by ColtonHoksbergen** [comment]
I'm not having any of the error messages unless I'm debugging and breakpoints delay my read, but my ReadAsync() always caps out at 32256 bytes, resulting in a corrupted image if the image was larger than this. Anyone have a solution for this?

**Comment by haas-daniel** [comment]
> I'm not having any of the error messages unless I'm debugging and breakpoints delay my read, but my ReadAsync() always caps out at 32256 bytes, resulting in a corrupted image if the image was larger than this. Anyone have a solution for this?

I think your problem is the SignalR message size limit @ColtonHoksbergen. This defaults to 32KB.

https://learn.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-7.0&pivots=server#signalr-message-size-limit

**Comment by ColtonHoksbergen** [comment]
> > I'm not having any of the error messages unless I'm debugging and breakpoints delay my read, but my ReadAsync() always caps out at 32256 bytes, resulting in a corrupted image if the image was larger than this. Anyone have a solution for this?
> 
> I think your problem is the SignalR message size limit @ColtonHoksbergen. This defaults to 32KB.
> 
> https://learn.microsoft.com/en-us/aspnet/core/blazor/file-uploads?view=aspnetcore-7.0&pivots=server#signalr-message-size-limit

Thank you very much, that was the exact issue. I know this is getting unrelated, but while my stream can now accept larger file types, the position is always 32256 regardless of if the stream is larger, and it is not seekable. Do you know what I might be doing wrong?

**Comment by alpisala02** [comment]
> Configure<HubOptions>(options => { options.DisableImplicitFromServicesParameters = true; });

Thank you so much! It worked perfectly.

**Comment by SammyGCh** [comment]
Hi, I faced to this issue and the @12evenant answer resolved it.

**Comment by Liero** [comment]
Please, upvote the original issue, so that it will be prioritized

**Comment by arielmoraes** [comment]
I'm also facing that issue, but in my case I'm reading and writing manually to another stream to have a progress callback.

When the streaming operation finishes with success there is no problem, but if any Exception is thrown the app hangs. 

The method responsible for the streaming operation is asynchronous, so the root cause might be because the OpenReadStream method is not closing the stream after a failure in the current async flow.

**Comment by philnwoha** [comment]
> await stream.ReadAsync(fileBytes);



> Hi All, In my case I was able to solve this by closing the stream.
> 
> ```
>     private async Task<bool> Upload(IBrowserFile file)
>     {
>         var fileBytes = new byte[file.Size];
>         var stream = file.OpenReadStream();
>         try
>         {
>             await stream.ReadAsync(fileBytes);
>             //do something
>             return true;
>         }
>         catch (Exception ex)
>         {
>             //do something
>             return false;
>         }
>         finally
>         {
>             stream.Close();
>         }
>     }
> ```

This worked for me.

**Comment by dadazxy** [comment]
> 我有同样的问题，并且解决方法不起作用。还有其他解决方案吗？

有解决吗

**Comment by AdamJonsson** [comment]
These workarounds worked for me: 
https://github.com/dotnet/aspnetcore/issues/47301#issuecomment-1520676385

**Comment by nblumhardt** [comment]
`DisableImplicitFromServicesParameters` should probably be the default here. Otherwise, SignalR/the framework is making assumptions about container behavior that are only valid for the MEDI container.

Autofac (right back as far as the beginning of .NET Core) has always regarded `IEnumerable<T>` or `T[]` as resolvable, even if no implementations exist, because `IEnumerable<T>` is a zero-to-many dependency. Several other containers follow this model IIRC.

Unfortunate that this wasn't spotted before the original release of this API, not sure what can be done now, but it's little friction points like this that erode the value in having multiple independent implementations out there :-)

For example, if the default behavior can't be changed, could the implicit-from-services mechanism internally apply a filter to decide what types constitute valid "services"? Or ignore registered services that are enumerable and empty?

**Comment by nblumhardt** [comment]
How about - leave the default intact, but skip implicit injection of services that are arrays of structs, or generic collection types/interfaces where the closing type is a struct?

I'm not sure how many scenarios apart from `byte[]` are actually interesting, but I also can't think of many cases where collections of structs would come from DI. But, leaving the attribute as an explicit opt-in would cover anything we're missing.

Could be a bit tedious to get right, but I'm game to try assembling a PR if it'd help.

**Comment by erosnemesis** [comment]
> > await stream.ReadAsync(fileBytes);
> 
> > Hi All, In my case I was able to solve this by closing the stream.
> > ```
> >     private async Task<bool> Upload(IBrowserFile file)
> >     {
> >         var fileBytes = new byte[file.Size];
> >         var stream = file.OpenReadStream();
> >         try
> >         {
> >             await stream.ReadAsync(fileBytes);
> >             //do something
> >             return true;
> >         }
> >         catch (Exception ex)
> >         {
> >             //do something
> >             return false;
> >         }
> >         finally
> >         {
> >             stream.Close();
> >         }
> >     }
> > ```
> 
> This worked for me. Thanks @12evenant

Can confirm that closing the steam solves the issue.

**Comment by umartechboy** [comment]
I can confirm this too. Closing the stream solves a similar issue in my project too. 

**Scenario:**
MudBlazor installed
FileChanged is called by MudBlazor file selector which is placed inside a MudDialog, reading the stream into buffer works. The rest of the process triggers too, around 3-4 seconds later, the UI Updates from the code become unresponsive. 20-30 seconds later, Timeout exception is thrown.

It was so frustrating to find the root cause because all my tests indicated problems with MudBlazor MudDialog or the CPU intensive work I was doing on the loaded data (WhisperNet. I had to change the library from Whisper.net to WhisperNet, now I feel sorry for blaming Whisper.net)

**Resolve**
Close the IBrowserFile stream right after read is finished!! A dumb thing to forget, I'd say. Only a desktop app developer can make the mistake.

**Comment by DontEatRice** [comment]
Any news on this? Still happening when we try to upload multiple files

**Comment by TheAnachronism** [comment]
I run into the same issue when doing things parallel.
Although it doesn't actually happen _all the time_.  I haven't been able to figure out which conditions actually cause the exception to happen, which makes it all the more difficult trying to find it.
And like @DontEatRice has mentioned, closing the streams doesn't make a difference, as nothing is received in the first place.

**Comment by MackinnonBuck** [comment]
> Any news on this? Still happening when we try to upload multiple files. We do not use AutoFac. Scenario: server-side Blazor tries to upload more than one file to an external service using HttpClient (RestSharp as wrapper) using streams. When there is more than one file `System.TimeoutException: Did not receive any data in the allotted time.` is thrown. Closing streams is pointless when they are being read in parallel

@DontEatRice, would you be able to provide us with a repro that demonstrates the bug? The repro we've seen so far wasn't reading the stream to completion (see https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-1151415351), and others using Autofac mentioned that https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-1342540950 resolved the issue for them.

A repro that demonstrates a framework bug would be very valuable, so if you (or someone else) can provide a minimal repro that doesn't use Autofac and is hosted as a public GitHub repository, that would be much appreciated. Thanks!

**Comment by DontEatRice** [comment]
@MackinnonBuck

I have created the repo that demonstrates this bug: https://github.com/DontEatRice/BlazorFileUploadIssue
This repo has no third-party dependencies. I have also included a simple api service that saves files to disk.

First run the Web Api in the Api folder.

Go to the BlazorApp folder and run it too. On the left panel you will see the Upload Files tab. You will find sample files in the BlazorApp/Data folder. They have been created using the fsutil command. Select all of them to see the bug. Of course you can use your own files, but the requirement is that they must be at least 2MB each (approximately).

After trying to upload them, Blazor will throw an exception after some time.
You can upload any single file and it will work.

In net7.0 the exception stack trace was longer, something happened, and in net8.0 it is quite short.

Please message me if you have any questions.

**Comment by MackinnonBuck** [comment]
Thank you for the repro, @DontEatRice! It looks like this is exactly the same issue as #47301, and you can see a detailed explanation [here](https://github.com/dotnet/aspnetcore/issues/47301#issuecomment-1520676385) that shows what's going on.

The [code in your repro](https://github.com/DontEatRice/BlazorFileUploadIssue/blob/39c346bbd5ab494715bb862ca46dada8d5bf95da/BlazorApp/Pages/UploadHttp.razor#L25-L29) can be modified slightly to make this work:

```diff
+var memoryStream = new MemoryStream();
+var browserFileStream = file.OpenReadStream(maxAllowedSize: 15_000_000);
+await browserFileStream.CopyToAsync(memoryStream);
+memoryStream.Position = 0;

-var fileContent = new StreamContent(file.OpenReadStream(long.MaxValue));
+var fileContent = new StreamContent(memoryStream);
fileContent.Headers.ContentType = new MediaTypeHeaderValue(file.ContentType);
fileContent.Headers.ContentDisposition = new ContentDispositionHeaderValue("form-data") { Name = "\"Files\"", FileName = $"\"{file.Name}\"" };

content.Add(fileContent);
```

---
Within this single GitHub issue, we've seen multiple causes of this exception:
* The app failing to read the stream to completion after calling `IBrowserFile.OpenReadStream` (not a framework bug)
* Autofac injecting a `byte[]` argument for a hub method parameter (workaround [here](https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-1342540950))
* A deadlock caused by opening multiple `IBrowserFile`s for reading at the same time, but consuming them sequentially (framework bug, tracked by #47301).

So, it's not clear to me whether this issue tracks a unique framework bug. There's a lot of community activity here, but the root cause of the problem seems to vary widely. This probably indicates we should update our documentation to help developers diagnose this issue better.

@guardrex, would you be able to add a section to the file upload docs that explain some of the common causes of this exception?

Also, it looks like the sample in [this section](https://learn.microsoft.com/aspnet/core/blazor/file-uploads?view=aspnetcore-8.0#upload-files-to-a-server-with-server-side-rendering) might be susceptible to this problem. If that's the case, could we update that sample to take an approach similar to what I wrote above (in the diff)? Thanks!

**Comment by guardrex** [comment]
@MackinnonBuck ... This one came in, too, which I was waiting to discuss around 🦃 Day. I'll mention this issue to that dev.

https://github.com/dotnet/AspNetCore.Docs/issues/33272

**Comment by TheAnachronism** [comment]
I have a question about the code example. The documentation mentions that files shouldn't be copied in to a `MemoryStream`, but should be directly used for whatever happens to them next. (https://learn.microsoft.com/en-gb/aspnet/core/blazor/file-uploads?view=aspnetcore-8.0#file-uploads)

I've done this for my implementation and passed the stream from `OpenReadStream` to the `PutAsync` method of a MinioClient for an S3 upload.

**Comment by MackinnonBuck** [comment]
> The documentation mentions that files shouldn't be copied in to a MemoryStream, but should be directly used for whatever happens to them next.

That's true in most cases, but in the repro provided by @DontEatRice, the reading of each stream is controlled by the implementation of [`MultipartContent`](https://source.dot.net/#System.Net.Http/System/Net/Http/MultipartContent.cs,200), which only consumes each stream just before the request gets made. To avoid the deadlock, each stream needs to be completely consumed before the next call to `IBrowserFile.OpenReadStream()`. If you want to avoid reading the entire stream into memory, you could limit each request to only transmitting a single file.

Note that this workaround won't be required:
* If using WebAssembly interactivity, or
* When limiting uploads to one file at a time, or
* After #47301 gets fixed

**Comment by TheAnachronism** [comment]
Ah, thank you. Guess I'll change this until that issue is resolved.

**Comment by MackinnonBuck** [comment]
I'm going to close this out now:
* We've updated our guidance in the docs to account for various possible causes of this exception that aren't framework bugs
* The remaining framework bug is being tracked by https://github.com/dotnet/aspnetcore/issues/47301

If others are experiencing this issue and the root cause is not one of those listed in [this comment](https://github.com/dotnet/aspnetcore/issues/38842#issuecomment-2286886600), please open a new issue with detailed reproduction steps and a minimal project that can be used to reproduce the issue (hosted as a public GitHub repository, not a zip file). Thanks!

**Comment by Wolfware2023** [comment]
Has anyone been able to reproduce this in a BUnit test?
I tried to reproduce the error in a test, but even when it takes forever (>60 seconds) the process does not get timed out.
